src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,345 @@
+                if (load.Any(l => l.ProvenBlockHeader == null))
+                {
+                    this.logger.LogTrace("(-)[PROVEN_BLOCK_HEADER_INFO_MISSING]");
+                    throw new ConfigurationException("Missing proven block header information, delete the data folder and re-download the chain");
+                }
+
+                foreach (StakeItem stakeItem in load)
+                    this.stakeItemQueue.Enqueue(stakeItem);
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <inheritdoc />
+        public async Task<ProvenBlockHeader> GetAsync(uint256 blockId, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(blockId), blockId);
+
+            List<StakeItem> item = new List<StakeItem>
+            {
+                new StakeItem
+                {
+                    BlockId = blockId
+                }
+            };
+
+            await this.provenBlockHeaderRepository.GetAsync(item, cancellationToken).ConfigureAwait(false);
+
+            var provenBlockHeader = item.FirstOrDefault().ProvenBlockHeader;
+
+            if (provenBlockHeader != null)
+                this.logger.LogTrace("(-):*.{0}='{1}'", nameof(provenBlockHeader), provenBlockHeader);
+            else
+                this.logger.LogTrace("(-):null");
+
+            Guard.Assert(provenBlockHeader != null);
+
+            return provenBlockHeader;
+        }
+
+        /// <inheritdoc />
+        public async Task<uint256> GetTipHashAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            this.logger.LogTrace("()");
+
+            this.TipHash = await this.provenBlockHeaderRepository.GetTipHashAsync(cancellationToken).ConfigureAwait(false);
+
+            this.logger.LogTrace("(-)");
+
+            return this.TipHash;
+        }
+
+        /// <inheritdoc />
+        public async Task<ProvenBlockHeader> GetTipAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            this.logger.LogTrace("()");
+
+            this.TipHash = await this.GetTipHashAsync(cancellationToken);
+
+            this.logger.LogTrace("(-)");
+
+            return await this.GetAsync(this.TipHash).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc />
+        public void AddToPending(ChainedHeader chainedHeader, ProvenBlockHeader provenBlockHeader)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(provenBlockHeader), provenBlockHeader);
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeader), chainedHeader);
+
+            var item = new StakeItem
+            {
+                BlockId = chainedHeader.HashBlock,
+                Height = chainedHeader.Height,
+                ProvenBlockHeader = provenBlockHeader,
+                InStore = false
+            };
+
+            this.stakeItemQueue.Enqueue(item);
```

Comment:
I hope we aren't inadvertently introducing a similar race condition here to what we had in the regular block store. (See #2162 for a description of that issue, #2207 for the fix)

Reply:
Thanks for the heads up!

Currently `ProvenBlockHeaderStore` doesn't include logic for caching (which is part of #2143).  

When the caching object work starts - which will use a dictionary, like the code below.  I'll make sure I take the race condition into consideration :  

https://github.com/stratisproject/StratisBitcoinFullNode/blob/0f1b572df682235b998ad542c5ff857a63a9f6f4/src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs#L79).